src/Panel/Panel.tsx

```diff
@@ -0,0 +1,237 @@
+    return this._softLoadingActive || (softLoading && softLoadingDelay === 0); // optimization for the default delay === 0;
+  }
+
+  componentDidMount() {
+    this.updateSoftLoadingState(false);
+  }
+
+  componentDidUpdate(oldProps: PanelDefaultedProps) {
+    this.updateSoftLoadingState(oldProps.softLoading);
+  }
+
+  componentWillUnmount() {
+    if (this._softLoadingTimer) {
+      clearTimeout(this._softLoadingTimer);
+    }
+  }
+
+  toggleExpanded = () => {
+    if (this.props.header && this.props.header.collapse) {
+      const { isCollapsed, onExpand, onCollapse } = this.props.header.collapse;
+      if (isCollapsed) {
+        onExpand();
+      } else {
+        onCollapse();
+      }
+    }
+  };
+
+  getStyle = () => {
+    const style = {};
+    if (this.props.clearMargin) {
+      const marginToClear = `margin${capitalize(this.props.clearMargin)}`;
+      style[marginToClear] = 0;
+    }
+    return {
+      ...this.props.style,
+      ...style
+    };
+  };
+
+  templateSoftLoading = ({ softLoading, isExpanded }: {
+    softLoading: boolean,
+    isExpanded: boolean
+  }) => {
+    return softLoading && isExpanded ? (
+      <div className='panel-soft-loader'>
+        <div className='loader gradient' />
+        <div className='loader' />
+      </div>
+    ) : null;
+  };
+
+  templateHeader = ({ header, isExpanded, toggleExpanded }: {
+    header?: PanelProps.Header,
+    isExpanded: boolean,
+    toggleExpanded: () => void
+  }) => {
+    return (
+      header ?
+        <PanelHeader
+          title={header.hideTitleWhenExpanded && isExpanded ? undefined : header.title}
+          size={header.size}
+          content={header.content}
+          menu={header.menu}
+          collapse={header.collapse ? {
+            direction: header.collapse.direction,
+            isExpanded,
+            onToggleExpanded: toggleExpanded
+          } : undefined}
+        />
+    : null );
+  };
+
+  templateExpandedContent = ({ children, loading }: {
+    children: React.ReactNode,
+    loading: boolean
+  }) => {
+    return (
+      <FlexView className='panel-content' column grow style={{ position: 'relative' }}>
+        {children}
+        {loading && <LoadingSpinner />}
+      </FlexView>
+    );
+  };
+
+  templateCollapsedContent = ({ header, verticalDirection }: {
+    header?: PanelProps.Header,
+    verticalDirection?: boolean
+  }) => {
+    return (
+      (!verticalDirection && header && header.title) ?
+        <FlexView className='panel-content' column grow>
+          <FlexView grow className='panel-content-title'>
+            {header.title}
+          </FlexView>
+        </FlexView>
+        : null
+    );
+  };
+
+  render() {
+    const { header, children, loading, softLoading: _softLoading, softLoadingDelay, type, className: _className, dark } = this.props as PanelDefaultedProps;
+    const collapsable = !!header && !!header.collapse;
+    const isExpanded = !collapsable || !header!.collapse!.isCollapsed;
```

Comment:
all these `!` on `header` and `header.collapse` are "very brutti", but we're sure they're safe since we're checking `collapsable`, that's equivalent to check  both `header` and `header.collapse`